### PR TITLE
Support glm4.1v and glm4.5v

### LINKED ISSUE #7993: [Feature] Any plan to support glm-4.1v-thinking
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

I hope this message finds you well. I would like to kindly request your support for integrating the glm-4.1v-thinking model into the SGLang framework. This model offers significant capabilities for advanced reasoning tasks, and its inclusion would be greatly beneficial for the community. Thank you for your continued efforts and support.

### Related resources

https://github.com/THUDM/GLM-4.1V-Thinking

### PR DESCRIPTION
**Pull request by byjiang1996**
## Motivation

Close https://github.com/sgl-project/sglang/issues/7993. 

Adapted from https://github.com/sgl-project/sglang/pull/8015 and depend on https://github.com/huggingface/transformers/pull/39805

Covers https://github.com/sgl-project/sglang/pull/8948

## Modifications

This pull request introduces support for the GLM-4V model within the sglang framework. It includes the necessary model files, configuration adjustments, and a new conversation template to facilitate interaction with the GLM-4V model. Additionally, it addresses potential issues with MRoPE section mismatches and enhances the flexibility of the get_rope_index function to accommodate various model configurations.

**Model Support**: Adds support for the GLM-4V model, including necessary configurations and model files.
Conversation Template: Registers a new conversation template specifically for GLM-4V, defining system messages, roles, and separators.
**MRoPE Section Adjustment**: Implements auto-correction for MRoPE section mismatches in rotary embeddings, ensuring compatibility with different model configurations.
**Vision Start Token ID**: Modifies the get_rope_index function to accept a list of vision start token IDs, accommodating models with multiple vision start tokens.
**Conditional Logic**: Adds conditional logic to handle specific model types like 'glm4v' in the get_rope_index function.

## Accuracy Test

GLM4.1v and GLM4.5v UT pass: `single image`, `multi image`, `video image` and `video`

## Benchmark & Profiling

GLM4.1v mmmu = 0.701 - aligns with the tech report https://huggingface.co/zai-org/GLM-4.1V-9B-Thinking 

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [x] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/function_call/glm4_moe_detector.py [hunk 1]
```diff
@@ -158,7 +158,7 @@ def build_ebnf(self, tools: List[Tool]):
             individual_call_end_token=self.eot_token,
             tool_call_separator="\\n",
             function_format="xml",
-            call_rule_fmt='"{name}" "\\n" {arguments_rule} "\\n"',
+            call_rule_fmt='"{name}" "\\n" ( {arguments_rule} "\\n" )?',
```
Comment (JustinTong0323) [inline python/sglang/srt/function_call/glm4_moe_detector.py:161]: CC @CatherineSue I am not sure about here, may you take a look?

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1153,6 +1184,206 @@ def get_rope_index(
...
+    def get_rope_index_glm4v(
+        input_ids: torch.Tensor,
+        hf_config: Any,
+        image_grid_thw: Union[list[list[int]], torch.Tensor],
+        video_grid_thw: Union[list[list[int]], torch.Tensor],
+        attention_mask: torch.Tensor,
+        **kwargs,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """Get mrope input positions and delta value for GLM4V."""
+        # device = input_ids.device
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/rotary_embedding.py]: nit: clean up

File: test/srt/test_vision_openai_server_b.py [hunk 3]
```diff
@@ -301,6 +301,24 @@ def setUpClass(cls):
...
+    @classmethod
+    def setUpClass(cls):
+        cls.model = "zai-org/GLM-4.1V-9B-Thinking"
+        cls.base_url = DEFAULT_URL_FOR_TEST
+        cls.api_key = "sk-123456"
+        cls.process = popen_launch_server(
+            cls.model,
+            cls.base_url,
+            timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
+            other_args=["--trust-remote-code", "--mem-fraction-static", "0.68"],
```
Comment (JustinTong0323) [inline test/srt/test_vision_openai_server_b.py]: nit: set max cuda graph bs to 4 for quicker CI

File: test/srt/test_vision_openai_server_common.py [hunk 4]
```diff
@@ -308,11 +317,12 @@ def test_video_images_chat_completion(self):
             "iPod" in video_response
             or "device" in video_response
             or "microphone" in video_response
+            or "rectangular" in video_response
         ), f"""
         ====================== video_response =====================
         {video_response}
         ===========================================================
-        should contain 'iPod' or 'device' or 'microphone'
+        should contain 'iPod' or 'device' or 'microphone' or 'rectangular'
```
Comment (JustinTong0323) [inline test/srt/test_vision_openai_server_common.py]: what is rectangular? 🤔

File: python/sglang/srt/layers/rotary_embedding.py (hunk 2 above)
Comment (byjiang1996) [reply python/sglang/srt/layers/rotary_embedding.py]: done

File: test/srt/test_vision_openai_server_b.py (hunk 3 above)
Comment (byjiang1996) [reply test/srt/test_vision_openai_server_b.py]: good point! done

File: test/srt/test_vision_openai_server_common.py (hunk 4 above)
Comment (byjiang1996) [reply test/srt/test_vision_openai_server_common.py]: Just had a try. it seems no longer needed. Removed.

File: benchmark/mmmu/eval_utils.py [hunk 5]
```diff
@@ -226,7 +233,7 @@ def process_sample(i, sample):
 
 
 def get_sampling_params(eval_args):
-    max_new_tokens = 30
```
Comment (mickqian) [inline benchmark/mmmu/eval_utils.py]: better to be careful with this change, unless for reasoning models

Comment (mickqian) [reply benchmark/mmmu/eval_utils.py]: can you set this with cmd argument like `reasoning`?

Comment (byjiang1996) [reply benchmark/mmmu/eval_utils.py]: Thanks for raising this Mick!

I changed the default back to 30 and add how-to-change-max-token-value in the README. Because xhs vlm needs the same functionality, I extracted mmmu changes to another PR here: https://github.com/sgl-project/sglang/pull/8948

File: docs/supported_models/multimodal_language_models.md [hunk 6]
```diff
@@ -39,3 +39,4 @@ in the GitHub search bar.
 | **Mistral-Small-3.1-24B**  | `mistralai/Mistral-Small-3.1-24B-Instruct-2503` | `mistral`   | Mistral 3.1 is a multimodal model that can generate text from text or images input. It also supports tool calling and structured output. |
 | **Phi-4-multimodal-instruct**  | `microsoft/Phi-4-multimodal-instruct` | `phi-4-mm`   | Phi-4-multimodal-instruct is the multimodal variant of the Phi-4-mini model, enhanced with LoRA for improved multimodal capabilities. It supports text, vision and audio modalities in SGLang. |
 | **MiMo-VL** (7B)           | `XiaomiMiMo/MiMo-VL-7B-RL`                 | `mimo-vl`        | Xiaomi's compact yet powerful vision-language model featuring a native resolution ViT encoder for fine-grained visual details, an MLP projector for cross-modal alignment, and the MiMo-7B language model optimized for complex reasoning tasks. |
+| **GLM-4.5V** (106B)           | `zai-org/GLM-4.5V`                   | `glm-4v`         | ZhipuAI's vision-language extension of GLM                                                                                                                                                                                                      |
```
Comment (JustinTong0323) [inline docs/supported_models/multimodal_language_models.md]: maybe add 4.1V as well

File: python/sglang/srt/function_call/glm4_moe_detector.py (hunk 1 above)
Comment (byjiang1996) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: This is the bug reported from GLM team. For this request, the ebnf schema is incorrect due to empty function parameter field
```
"tools": [
      {
        "type": "function",
        "function": {
          "name": "request_function",
          "description": "",
          "parameters": {
            "type": "object",
            "properties": {},
            "required": []
          }
        }
      }
    ],
    "tool_choice": {
      "type": "function",
      "function": {
        "name": "request_function"
      }
    },
```

Personally a bit feeling this should be the right format:
- we have one extra `\n` at end of call_rule_fmt before this code change
- in addition, `arguments_rule` could've been optional fields
```
            call_rule_fmt='"{name}" ( "\\n" {arguments_rule} )?',
```

image

Comment (CatherineSue) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: what is the ebnf string for the above case?

File: python/sglang/srt/function_call/ebnf_composer.py [hunk 7]
```diff
@@ -316,6 +316,8 @@ def build_ebnf(
 
             combined_args = "".join(rule_parts)
             arguments_rule = args_template.format(arg_rules=combined_args)
+            if len(arguments_rule) == 0:
```
Comment (CatherineSue) [inline python/sglang/srt/function_call/ebnf_composer.py]: Can we check rule_parts instead? It is easier to read.

File: python/sglang/srt/function_call/glm4_moe_detector.py (hunk 1 above)
Comment (CatherineSue) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: Do we need to make arguments optional once we fix the ebnf_composer?

Comment (CatherineSue) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: > we have one extra \n at end of call_rule_fmt before this code change

I am not following this. What do you mean by before this code change?

File: python/sglang/srt/function_call/ebnf_composer.py (hunk 7 above)
Comment (CatherineSue) [reply python/sglang/srt/function_call/ebnf_composer.py]: or
```suggestion
            arguments_rule = arguments_rule or '""'
```

Comment (CatherineSue) [reply python/sglang/srt/function_call/ebnf_composer.py]: Can we also add a test case for this at https://github.com/sgl-project/sglang/blob/main/test/srt/test_function_call_parser.py#L627？

File: python/sglang/srt/function_call/glm4_moe_detector.py (hunk 1 above)
Comment (byjiang1996) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: For this tool
```
class TestEBNFGeneration(unittest.TestCase):
    def setUp(self):
        # Create sample tools for testing
        self.tools = [
            Tool(
                type="function",
                function=Function(
                    name="get_weather",
                    description="Get weather information",
                    parameters={
                        "properties": {},
                        "required": [],
                    },
                ),
            ),
            Tool(
                type="function",
                function=Function(
                    name="search",
                    description="Search for information",
                    parameters={
                        "properties": {},
                        "required": [],
                    },
                ),
            ),
        ]
```

Before fix
```
root ::= "<tool_call>" function_call "</tool_call>" ( "\n" "<tool_call>" function_call "</tool_call>" )*
function_call ::= call_get_weather | call_search
call_get_weather ::= "get_weather" "\n" arguments_get_weather "\n" 
arguments_get_weather ::= ""
call_search ::= "search" "\n" arguments_search "\n" 
arguments_search ::= ""

        xml ::= xml_element | xml_text
        xml_element ::= basic_string | basic_number | basic_boolean | basic_null | basic_array | basic_object
        xml_text ::= [^<>]*
        basic_any ::= basic_number | basic_string | basic_boolean | basic_null | basic_array | basic_object
        basic_boolean ::= "true" | "false"
        basic_null ::= "null"
    
        basic_string ::= (([\"] basic_string_1 [\"]))
        basic_string_1 ::= "" | [^"\\\x00-\x1F] basic_string_1 | "\\" escape basic_string_1
        escape ::= ["\\/bfnrt] | "u" [A-Fa-f0-9]{4}
        basic_integer ::= ("0" | "-"? [1-9] [0-9]*) ".0"?
        basic_number ::= ("0" | "-"? [1-9] [0-9]*) ("." [0-9]+)? ([eE] [+-]? [0-9]+)?
        basic_array ::= "[" ("" | ws basic_any (ws "," ws basic_any)*) ws "]"
        basic_object ::= "{" ("" | ws basic_string ws ":" ws basic_any ( ws "," ws basic_string ws ":" ws basic_any)*) ws "}"
        ws ::= [ \n\t]*
```

After fix
```
root ::= "<tool_call>" function_call "</tool_call>" ( "\n" "<tool_call>" function_call "</tool_call>" )*
function_call ::= call_get_weather | call_search
call_get_weather ::= "get_weather" "\n" ( arguments_get_weather "\n" )?
arguments_get_weather ::= ""
call_search ::= "search" "\n" ( arguments_search "\n" )?
arguments_search ::= ""

        xml ::= xml_element | xml_text
        xml_element ::= basic_string | basic_number | basic_boolean | basic_null | basic_array | basic_object
        xml_text ::= [^<>]*
        basic_any ::= basic_number | basic_string | basic_boolean | basic_null | basic_array | basic_object
        basic_boolean ::= "true" | "false"
        basic_null ::= "null"
    
        basic_string ::= (([\"] basic_string_1 [\"]))
        basic_string_1 ::= "" | [^"\\\x00-\x1F] basic_string_1 | "\\" escape basic_string_1
        escape ::= ["\\/bfnrt] | "u" [A-Fa-f0-9]{4}
        basic_integer ::= ("0" | "-"? [1-9] [0-9]*) ".0"?
        basic_number ::= ("0" | "-"? [1-9] [0-9]*) ("." [0-9]+)? ([eE] [+-]? [0-9]+)?
        basic_array ::= "[" ("" | ws basic_any (ws "," ws basic_any)*) ws "]"
        basic_object ::= "{" ("" | ws basic_string ws ":" ws basic_any ( ws "," ws basic_string ws ":" ws basic_any)*) ws "}"
        ws ::= [ \n\t]*
```

Comment (byjiang1996) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: I just pushed new version of ebnf to the branch. Please help take a look @CatherineSue . thank you!

File: python/sglang/srt/function_call/ebnf_composer.py (hunk 7 above)
Comment (byjiang1996) [reply python/sglang/srt/function_call/ebnf_composer.py]: Yes. Added.

File: python/sglang/srt/function_call/glm4_moe_detector.py (hunk 1 above)
Comment (CatherineSue) [reply python/sglang/srt/function_call/glm4_moe_detector.py:161]: LGTM, thanks for the reply!

**Review by CatherineSue — APPROVED** [review]

File: python/sglang/srt/layers/rotary_embedding.py [hunk 8]
```diff
@@ -1153,6 +1184,204 @@ def get_rope_index(
...
             return position_ids, mrope_position_deltas
 
+    # Adapted from https://github.com/vllm-project/vllm/blob/3779eb8c81449b924a23457fc77e45a0e6171178/vllm/model_executor/layers/rotary_embedding.py#L1120
+    @staticmethod
+    def get_rope_index_glm4v(
+        input_ids: torch.Tensor,
+        hf_config: Any,
+        image_grid_thw: Union[list[list[int]], torch.Tensor],
+        video_grid_thw: Union[list[list[int]], torch.Tensor],
+        attention_mask: torch.Tensor,
```
Comment (mickqian) [inline python/sglang/srt/layers/rotary_embedding.py:1194]: `attention_mask` can be improved, see `get_rope_index`
